specs/FrameProcesssInfo.md

```diff
@@ -0,0 +1,503 @@
+{
+    UINT32 frameId = 0;
+    UINT32 parentFrameId = 0;
+    UINT32 mainFrameId = 0;
+    UINT32 firstLevelFrameId = 0;
+    std::wstring type = L"other child frame";
+    wil::unique_cotaskmem_string nameRaw;
+    wil::unique_cotaskmem_string sourceRaw;
+    COREWEBVIEW2_FRAME_KIND frameKind = COREWEBVIEW2_FRAME_KIND_OTHER;
+
+    CHECK_FAILURE(frameInfo->get_Name(&nameRaw));
+    std::wstring name = ((std::wstring)(nameRaw.get())).empty() ? L"none" : nameRaw.get();
+    CHECK_FAILURE(frameInfo->get_Source(&sourceRaw));
+    std::wstring source = ((std::wstring)(sourceRaw.get())).empty() ? L"none" : sourceRaw.get();
+
+    wil::com_ptr<ICoreWebView2FrameInfo2> frameInfo2;
+    CHECK_FAILURE(frameInfo->QueryInterface(IID_PPV_ARGS(&frameInfo2)));
+    frameInfo2->get_FrameId(&frameId);
+    frameInfo2->get_FrameKind(&frameKind);
+
+    wil::com_ptr<ICoreWebView2FrameInfo> parentFrameInfo;
+    CHECK_FAILURE(frameInfo2->get_ParentFrameInfo(&parentFrameInfo));
+    if (parentFrameInfo)
+    {
+        CHECK_FAILURE(parentFrameInfo->QueryInterface(IID_PPV_ARGS(&frameInfo2)));
+        CHECK_FAILURE(frameInfo2->get_FrameId(&parentFrameId));
+    }
+
+    wil::com_ptr<ICoreWebView2FrameInfo> mainFrameInfo = GetAncestorMainFrameInfo(frameInfo);
+    if (mainFrameInfo == frameInfo)
+    {
+        type = L"main frame";
+    }
+    CHECK_FAILURE(mainFrameInfo->QueryInterface(IID_PPV_ARGS(&frameInfo2)));
+    CHECK_FAILURE(frameInfo2->get_FrameId(&mainFrameId));
+
+    wil::com_ptr<ICoreWebView2FrameInfo> firstLevelFrameInfo =
+        GetAncestorFirstLevelFrameInfo(frameInfo);
+    if (firstLevelFrameInfo == frameInfo)
+    {
+        type = L"first level frame";
+    }
+    if (firstLevelFrameInfo)
+    {
+        CHECK_FAILURE(
+            firstLevelFrameInfo->QueryInterface(IID_PPV_ARGS(&frameInfo2)));
+        CHECK_FAILURE(frameInfo2->get_FrameId(&firstLevelFrameId));
+    }
+
+    result << L"{frame name:" << name << L" | frame Id:" << std::to_wstring(frameId)
+           << L" | parent frame Id:"
+           << ((parentFrameId == 0) ? L"none" : std::to_wstring(parentFrameId))
+           << L" | frame type:" << type << L"\n"
+           << L" | ancestor main frame Id:" << std::to_wstring(mainFrameId)
+           << L" | ancestor first level frame Id:"
+           << ((firstLevelFrameId == 0) ? L"none" : std::to_wstring(firstLevelFrameId)) << L"\n"
+           << L" | frame kind:" << FrameKindToString(frameKind) << L"\n"
+           << L" | frame source:" << source << L"}," << std::endl;
+}
+
+// Get a string for the frame kind enum value.
+std::wstring ProcessComponent::FrameKindToString(const COREWEBVIEW2_FRAME_KIND kind)
+{
+    switch (kind)
+    {                                                                
+    case kindValue:   
+        return L#kindValue;
+
+        KIND_ENTRY(COREWEBVIEW2_FRAME_KIND_MAIN_FRAME);
+        KIND_ENTRY(COREWEBVIEW2_FRAME_KIND_IFRAME);

```

Comment:
Naming: I thought we'd decided at some point to consistently use "frame" rather than "iframe" because no one remembers framesets anyway.